qiskit/algorithms/amplitude_amplifiers/amplitude_amplifier.py

```diff
@@ -43,9 +43,9 @@ class AmplitudeAmplifierResult(AlgorithmResult):
 
     def __init__(self) -> None:
         super().__init__()
-        self._top_measurement = None
+        self._top_measurement: Optional[str] = None
```

Comment:
In Python 3.10, there is a new syntax for this that is preferred: `str | None`. This can already be used in older Python versions by adding `from __future__ import annotations` at the top of the file.

However, it is up to the Qiskit maintainers to decide to use the newer syntax now, or stick to `Optional[str]`.

Reply:
(Whoops I made 2 almost identical comments. I thought my first comment wasn't included by GitHub.)

Reply:
I guess it is the question of code style in general. I haven't seen usage of `|` instead of `Union` nowhere in codebase, so I stick to whatever already used.

Reply:
The existing annotation in the codebase was likely added before this was even a thing. From [PEP 604](https://peps.python.org/pep-0604/), it is the recommended syntax.

Reply:
I found evidence of ` | ` usage in the current qiskit-terra: https://github.com/Qiskit/qiskit-terra/blob/86fb555871f4811a842380dbf0d54dcda34a4841/qiskit/primitives/base_estimator.py#L112-L115.

Reply:
Yes, and we have been refactoring algorithms to use primitives and have been switching over to this style. With newer code we have been using that in preference to Union and Optional for the algorithms. The algorithms code come over to Terra from Aqua where it had been developed and would have been done with the older style annotations since the code was done some time ago.

FYI I will note that as part of the update of algorithms to primitives `'linear_solvers` (HHL) and `factorizers` (Shor) are now deprecated. The `minimum_eigen_solvers` and `eigen_solvers` in these locations are pending deprecation and will be deprecated and removed in the future. They are replaced by primitive based versions thereof in folders `minimum_eigensolvers` and `eigensolvers` (no underscore between eigen and solvers). `evolvers` are similarly replaced by `time_evolvers`. Amplitude amplifiers, amplitude estimators and phase estimators were updated in situ. Also `gradients` and `state_fidelities` have been added. This PR covered the algorithms folder - the above represents quite a bit of fairly recent change!